st3/sublime_lib/command_input.py

```diff
@@ -0,0 +1,70 @@
+from sublime_plugin import CommandInputHandler, Command, ListInputHandler, TextInputHandler
+
+from ._compat.typing import Callable, Generator, Optional, Union
+from functools import wraps
+
+FancyInputHandler = Union['GenericListInputHandler', 'GenericTextInputHandler']
+GeneratorType = Generator[FancyInputHandler, object, None]
+
+
+__all__ = ['fancy_input', 'GenericListInputHandler', 'GenericTextInputHandler']
+
+
+def fancy_input(
+    input_function: Callable[[Command, dict], GeneratorType],
+) -> Callable[[Command, dict], Optional[FancyInputHandler]]:
+    @wraps(input_function)
+    def wrapped(self: Command, args: dict) -> Optional[FancyInputHandler]:
+        return _next_input(input_function(self, args), None)
+
+    return wrapped
+
+
+def _next_input(gen: GeneratorType, value: object) -> Optional[FancyInputHandler]:
+    try:
+        next_input_handler = gen.send(value)
+        next_input_handler._gen = gen
+        return next_input_handler
+    except StopIteration:
+        return None
+
+
+class GenericListInputHandler(ListInputHandler):
+    _gen = None  # type: GeneratorType
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        items: list
```

Comment:
Other params we should look into supporting (and decide whether they are too advanced or not):

- [ ] placeholder
- [ ] initial_text
- [ ] initial_selection
- [ ] validate
- [ ] preview
- [ ] description
- [x] ~~confirm~~ (equivalent to the sent value into the generator)
- [ ] cancel

For cancel we could throw an exception to the generator.